### PR TITLE
Feature/#61 colav

### PR DESCRIPTION
**Pull request by oskarhelgerud890**
Implemented functionality for COLAV, works in simulations
Implemented a flipper between LOS and COLAV
Needs testing on hardware

### CODE REVIEW

File: motion/colav/scripts/VOMATH.py [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+#         constraints.append(cp.norm(velocity - obstacle_positions[i]) >= obstacle_radii[i])
+#     constraints.append(cp.norm(velocity) == current_speed)
+#     constraints.append((velocity @ current_velocity) / current_speed >= np.cos(theta_max))
+    
+#     # Define and solve problem
+#     problem = cp.Problem(obj, constraints)
+#     problem.solve()
+    
+#     # Return new velocity vector
+#     return velocity.value
```
Comment (theBadMusician) [inline motion/colav/scripts/VOMATH.py:231]: Remove is code is not going to be used

File: motion/colav/scripts/VOMATH.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+        if self.enable:
+            # for all obstacles:
+            # check if could have collison, red, green yellow
+            # if green do nothing
+            # if yellow give new heading
+            # if red, stop!
+            pass 
+        # TODO: 
+        #implemen tmath
+        #implement landmar server
```
Comment (theBadMusician) [inline motion/colav/scripts/VOMATH.py:208]: Complete the function or explain why it does not do anything (a TODO?), or remove if it is not needed

File: motion/colav/scripts/VOMATH.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+#     problem = cp.Problem(obj, constraints)
+#     problem.solve()
+    
+#     # Return new velocity vector
+#     return velocity.value
+
+
+        
+
+if __name__ == "__main__":
```
Comment (theBadMusician) [inline motion/colav/scripts/VOMATH.py:236]: Remove comments if not to be used. What is this main supposed to do?

File: motion/colav/scripts/VO_controller_node.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    """
+    The velocity object controller node.
+    Changes the ASV velocity such that a collision cannot occur.
+
+    """
+    
+
+    def __init__(self):
+        rospy.init_node("colav",anonymous=True)
```
Comment (theBadMusician) [inline motion/colav/scripts/VO_controller_node.py:22]: anonymous = false

File: motion/colav/scripts/VO_controller_node.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+
+
+    def avoid_collision(self,trigger):
+
+
+        """
+
+        Avoids collison by continuously updating the desired velociy,
+        and sending it to the AUV's velocity controller. This is implemented as a service that
```
Comment (theBadMusician) [inline motion/colav/scripts/VO_controller_node.py:51]: ASV

Comment (theBadMusician) [inline motion/colav/scripts/VO_controller_node.py:51]: We do not have a velocity controller, but we have a surge controller (named speed controller), and also a heading controller

File: motion/colav/scripts/VO_controller_node.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        Avoids collison by continuously updating the desired velociy,
+        and sending it to the AUV's velocity controller. This is implemented as a service that 
+        will run as long a collision is possible.
+
+        Current implementation always puts the vessel behind the obstacle
+
+        """
+
+
+        VO = VelocityObstacle(None,self.obstacle,self.vessel) #Placeholder None
```
Comment (theBadMusician) [inline motion/colav/scripts/VO_controller_node.py:59]: What is the None?

File: motion/colav/scripts/VO_controller_node.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        while VO.check_if_collision():
```
Comment (theBadMusician) [inline motion/colav/scripts/VO_controller_node.py:60]: maybe change function name to is_collision, this way, the return value is true if collision and false if not collision. Makes code mroe readable

File: motion/colav/scripts/VO_controller_node.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        self.obstacle = data
+
+
+if __name__ == "__main__":
+    try:
+        node = VOControllerNode()
+
+        rospy.spin()
+    except rospy.ROSInterruptException:
+        pass
```
Comment (theBadMusician) [inline motion/colav/scripts/VO_controller_node.py:96]: do not need try and except, use while not rospy is shutdown instead.

File: motion/colav/scripts/VO_visualize.py
Comment (theBadMusician) [inline motion/colav/scripts/VO_visualize.py:1]: Add some comments in the code and remove commented out code .

File: motion/colav/scripts/VOMATH.py
Comment (theBadMusician) [inline motion/colav/scripts/VOMATH.py:1]: This file should just be a library python file that gets imported where needed, without any ROS code.

Comment (theBadMusician) [inline motion/colav/scripts/VOMATH.py:1]: rename to VO_math

File: motion/los_guidance/scripts/los_guidance_node.py [hunk 9]
```diff
@@ -228,12 +254,7 @@ def lookaheadBasedSteering(self):
...
 class LosPathFollowing(object):
     """
-    This is the ROS wrapper class for the LOS class.
-    Attributes:
-            _feedback	A vortex_msgs action that contains the distance to goal
-            _result		A vortex_msgs action, true if a goal is set within the
-                                    sphereof acceptance, false if not
-
+        self.reference_mod
     Nodes created:
```
Comment (theBadMusician) [inline motion/los_guidance/scripts/los_guidance_node.py:258]: Why has this been deleted? Revert maybe, and this is crazy, but heres my number, so call me maybe

File: motion/los_guidance/scripts/los_guidance_node.py [hunk 10]
```diff
@@ -269,6 +290,11 @@ def __init__(self):
         self.sub = rospy.Subscriber(
             "/pose_gt", Odometry, self.callback, queue_size=1
         )  # 20hz
+        #Services
+        #placeholder srv name
```
Comment (theBadMusician) [inline motion/los_guidance/scripts/los_guidance_node.py:294]: remove placeholder comment

File: motion/los_guidance/scripts/los_guidance_node.py [hunk 11]
```diff
@@ -335,6 +361,10 @@ def fixHeadingWrapping(self):
 
         return x_d
 
+ 
+
+    
+
```
Comment (theBadMusician) [inline motion/los_guidance/scripts/los_guidance_node.py:367]: remove whitespace

**Review by theBadMusician — CHANGES_REQUESTED** [review]
I am using code in VOMATH is fuunctional and that FSM code has been tested. Needs to be cleaned up and unnecessary code removed.

**Review by theBadMusician — CHANGES_REQUESTED** [review]
I am using code in VOMATH is fuunctional and that FSM code has been tested. Needs to be cleaned up and unnecessary code removed.

**Comment by EliasAarekol** [comment]
⢿⣿⣿⣿⣭⠹⠛⠛⠛⢿⣿⣿⣿⣿⡿⣿⠷⠶⠿⢻⣿⣛⣦⣙⠻⣿
⣿⣿⢿⣿⠏⠀⠀⡀⠀⠈⣿⢛⣽⣜⠯⣽⠀⠀⠀⠀⠙⢿⣷⣻⡀⢿
⠐⠛⢿⣾⣖⣤⡀⠀⢀⡰⠿⢷⣶⣿⡇⠻⣖⣒⣒⣶⣿⣿⡟⢙⣶⣮
⣤⠀⠀⠛⠻⠗⠿⠿⣯⡆⣿⣛⣿⡿⠿⠮⡶⠼⠟⠙⠊⠁⠀⠸⢣⣿
⣿⣷⡀⠀⠀⠀⠀⠠⠭⣍⡉⢩⣥⡤⠥⣤⡶⣒⠀⠀⠀⠀⠀⢰⣿⣿
⣿⣿⡽⡄⠀⠀⠀⢿⣿⣆⣿⣧⢡⣾⣿⡇⣾⣿⡇⠀⠀⠀⠀⣿⡇⠃
⣿⣿⣷⣻⣆⢄⠀⠈⠉⠉⠛⠛⠘⠛⠛⠛⠙⠛⠁⠀⠀⠀⠀⣿⡇⢸
⢞⣿⣿⣷⣝⣷⣝⠦⡀⠀⠀⠀⠀⠀⠀⠀⡀⢀⠀⠀⠀⠀⠀⠛⣿⠈
⣦⡑⠛⣟⢿⡿⣿⣷⣝⢧⡀⠀⠀⣶⣸⡇⣿⢸⣧⠀⠀⠀⠀⢸⡿⡆
⣿⣿⣷⣮⣭⣍⡛⠻⢿⣷⠿⣶⣶⣬⣬⣁⣉⣀⣀⣁⡤⢴⣺⣾⣽⡇

wrong branch hehe, see new pull req

**Pull request closed without merging**